### PR TITLE
introduce callForSeenUsers and countSeenUsers

### PR DESCRIPTION
**Pull request by MorrisJobke**
- downstream of https://github.com/owncloud/core/pull/26361

cc @rullzer @nickvergessen @schiessle 

I will create a ticket to check our codebase for similar usages.

### CODE REVIEW

**Comment by MorrisJobke** [comment]
cc @icewind1991

**Comment by rullzer** [comment]
I like it.
:+1:

File: lib/private/User/Manager.php [hunk 1]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+				'appid', $queryBuilder->createNamedParameter('login'))
+			)
+			->andWhere($queryBuilder->expr()->eq(
+				'configkey', $queryBuilder->createNamedParameter('lastLogin'))
+			)
+			->andWhere($queryBuilder->expr()->isNotNull('configvalue')
+			);
+
+		$query = $queryBuilder->execute();
+		return (int)$query->fetchColumn();
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: Missing `$query->closeCursor()`

File: lib/private/User/Manager.php [hunk 2]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+	 */
+	public function countSeenUsers() {
+		$queryBuilder = \OC::$server->getDatabaseConnection()->getQueryBuilder();
+		$queryBuilder->select($queryBuilder->createFunction('COUNT(*)'))
+			->from('preferences')
+			->where($queryBuilder->expr()->eq(
+				'appid', $queryBuilder->createNamedParameter('login'))
+			)
+			->andWhere($queryBuilder->expr()->eq(
+				'configkey', $queryBuilder->createNamedParameter('lastLogin'))
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: code style is odd in here

File: lib/private/User/Manager.php [hunk 3]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+			->andWhere($queryBuilder->expr()->isNotNull('configvalue')
+			);
+
+		$query = $queryBuilder->execute();
+		return (int)$query->fetchColumn();
+	}
+
+	/**
+	 * @param \Closure $callback
+	 * @param string $search
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: Can't see this

File: lib/private/User/Manager.php [hunk 4]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+	 * Getting all userIds that have a listLogin value requires checking the
+	 * value in php because on oracle you cannot use a clob in a where clause,
+	 * preventing us from doing a not null or length(value) > 0 check.
+	 * 
+	 * @param int $limit
+	 * @param int $offset
+	 * @return string[] with user ids
+	 */
+	private function getSeenUserIds($limit = null, $offset = null) {
+		$queryBuilder = \OC::$server->getDatabaseConnection()->getQueryBuilder();
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: DI?

File: lib/private/User/Manager.php [hunk 5]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+	}
+
+	/**
+	 * returns how many users have logged in once
+	 *
+	 * @return int
+	 * @since 9.2.0
+	 */
+	public function countSeenUsers() {
+		$queryBuilder = \OC::$server->getDatabaseConnection()->getQueryBuilder();
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: DI?

File: lib/private/User/Manager.php [hunk 6]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+		}
+		if ($offset !== null) {
+			$queryBuilder->setFirstResult($offset);
+		}
+		$query = $queryBuilder->execute();
+		$result = [];
+
+		while ($row = $query->fetch()) {
+			$result[] = $row['userid'];
+		}
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: Missing `$query->closeCursor()`

File: lib/private/User/Manager.php [hunk 7]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+
+		return $result;
+	}
+	/**
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: new line before this?

File: lib/public/IUserManager.php [hunk 8]
```diff
@@ -145,6 +145,21 @@ public function countUsers();
...
+	 * @since 9.2.0
+	 */
+	public function countSeenUsers();
+
+	/**
+	 * @param \Closure $callback
+	 * @param string $search
+	 * @since 9.2.0
+	 */
+	public function callForSeenUsers (\Closure $callback);
```
Comment (nickvergessen) [inline lib/public/IUserManager.php]: awkward space

File: lib/private/User/Manager.php [hunk 9]
```diff
@@ -344,30 +350,121 @@ public function countUsers() {
...
+		$query = $queryBuilder->execute();
+		return (int)$query->fetchColumn();
+	}
+
+	/**
+	 * @param \Closure $callback
+	 * @param string $search
+	 * @since 9.2.0
+	 */
+	public function callForSeenUsers (\Closure $callback) {
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: awkward space

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/public/IUserManager.php [hunk 10]
```diff
@@ -145,6 +145,21 @@ public function countUsers();
...
+	 * returns how many users have logged in once
+	 *
+	 * @return int
+	 * @since 9.2.0
+	 */
+	public function countSeenUsers();
+
+	/**
+	 * @param \Closure $callback
+	 * @param string $search
```
Comment (nickvergessen) [inline lib/public/IUserManager.php]: invalid doc

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/private/User/Manager.php (hunk 4 above)
Comment (rullzer) [inline lib/private/User/Manager.php]: Nah that will cause the DB to be initialized to early. So then installing fails. We should fix that at some point. But probabaly can't right now.

**Comment by rullzer** [comment]
I fixed most of the comments.
Made the IConfig always injected.

Injecting the DB here fails hard as it gets initialized to early then.

I say lets get this in. :+1:

**Comment by nickvergessen** [comment]
Fine by me, also tested with the announcement center
Failing tests are sign-off only.

👍

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**